packages/middleware-sdk-s3/src/region-redirect-middleware.ts

```diff
@@ -38,7 +38,7 @@ export function regionRedirectMiddleware(clientConfig: PreviouslyResolved): Init
         if (
           clientConfig.followRegionRedirects &&
           // err.name === "PermanentRedirect" && --> removing the error name check, as that allows for HEAD operations (which have the 301 status code, but not the same error name) to be covered for region redirection as well
-          err?.$metadata?.httpStatusCode === 301
+          (err?.$metadata?.httpStatusCode === 301 || err?.$metadata?.httpStatusCode === 400)
```

Comment:
status code 400 for S3 should correspond to the error code `IllegalLocationConstraintException`